### ISSUE TITLE
[Feature] [Design] RL Integration: Batch-Invariant Diffusion Rollouts

### ISSUE THREAD

**Issue by Ronnie-Rui** [vllm-project/vllm-omni#5512]
## Problem

> Roadmap item: [vLLM-Omni #4901](https://github.com/vllm-project/vllm-omni/issues/4901)
> Design context: [#4864](https://github.com/vllm-project/vllm-omni/issues/4864), [#3435](https://github.com/vllm-project/vllm-omni/issues/3435)

RL rollout workers submit many requests concurrently, so the same logical request — same prompt, same seed — can land at a different batch index or next to different neighbors on each run. Because GPU floating-point addition is not associative, changing batch composition changes how reductions are partitioned, which can change the last few bits of the output.

Ordinary inference never notices. RL does, because rollout output is reused as training evidence: a last-bit difference in a logprob feeds into reward ranking, then advantage, then the gradient. What starts as a rounding difference ends up as a different training run.

So what we need is:

```text
output(A alone at BS1) == output(A inside a real BSN batch)
```

Bitwise, and holding when A changes batch index, when its neighbors change, and when the batch splits differently.

Two weaker properties are easy to mistake for this. Same-seed reruns matching is necessary but says nothing about batching. And running one request at a time (`max_num_seqs=1`) does give stable output, but by avoiding co-batching rather than surviving it — useful as a control baseline, nothing more. The practical risk is a system that quietly falls back to serialized execution and reports success, delivering the weaker property while claiming the stronger one.

## Dependency direction

```mermaid
flowchart LR
    A[VeRL-Omni<br/>explicit seeds, fixed-recipe E2E]
    B[vLLM-Omni<br/>request contract + capability gate]
    C[native vLLM<br/>batch-invariant kernels and runtime]
    A --> B --> C
```

vLLM-Omni reuses native vLLM batch-invariance facilities rather than reimplementing kernels, and must never depend on the training framework. The training side consumes the capability; it does not define it.

## Design: three enforcement layers

Three separate things have to be true at once, and they happen at different moments in a request's life, so they cannot be collapsed into a single switch.

```mermaid
flowchart TD
    A[Request submitted] --> B{BIC mode enabled?}
    B -- No --> C[Existing default behavior, unchanged]
    B -- Yes --> D[Layer 1 · RNG identity<br/>request boundary]
    D --> E[Layer 2 · Capability gate<br/>engine construction + each request]
    E -- Unsupported --> F[Fail closed, list every offending field]
    E -- Supported --> G[Layer 3 · Runtime bootstrap<br/>before model construction]
    G --> H[Real co-batched forward]
```

### Layer 1 — Make the randomness reproducible

An explicit integer seed is the only RNG identity that survives a process boundary. When a request crosses into a worker, a `torch.Generator` cannot be meaningfully serialized, so it degrades to `generator.initial_seed()`. That returns the seed the generator was *created* with, not its current position in the stream — so a caller-side generator that has already been drawn from reconstructs, in the worker, at a different point. Same identity on paper, different numbers in practice.

So BIC mode requires an explicit integer seed and rejects generator-only input instead of approximating it. Approximating means silently handing back a different set of random numbers, which is worse than an error.

### Layer 2 — Refuse configurations we have not proven

Batch invariance is a property of a specific combination — pipeline, hardware, dtype, parallelism, feature flags — not of the code in general. Native vLLM's own implementation illustrates this: on SM80 it installs Triton matmul overrides, while on SM90+ it instead disables cuBLAS split-k, because the source of batch variance differs by architecture. A result measured on one combination therefore says nothing about another, and the gate exists to encode which combination we actually measured.

It checks in two places, against different risks:

**At engine construction**, against the configuration — pipeline class, dtype, parallelism degrees, execution mode, feature flags. These cannot change afterwards, so one check at startup suffices. This catches an engine that should never have started.

**At every request submission**, against the request — shapes, guidance, externally injected tensors, per-request LoRA. This catches a request that exceeds the certified range even though the engine itself is configured correctly.

Two rules about how it rejects:

**Name every unsupported field at once**, not just the first one. Fixing a configuration one round-trip at a time is expensive, and stopping at the first error suggests the rest are fine.

**Never degrade quietly.** A rejected request must not fall back to serialized execution and still be reported as batch-invariant — that delivers the weaker property while claiming the stronger one, exactly the failure described above. A caller who wants serialized execution should ask for it explicitly.

### Layer 3 — Turn it on early enough

`init_batch_invariance()` is not a mode flag that the runtime consults per operation. It performs three kinds of one-time, process-wide mutation, and each one binds to state that later stages capture:

1. **Operator replacement.** It installs `torch.library.Library("aten", "IMPL")` overrides for `mm`, `addmm`, `matmul`, `linear`, `softmax`, `mean.dim`, and `bmm`, and rebinds `torch.bmm` itself. Any code that already resolved those symbols keeps the old implementation.
2. **Environment variables.** It sets `CUBLAS_WORKSPACE_CONFIG`, a full set of NCCL determinism variables (`NCCL_ALGO=allreduce:tree`, `NCCL_MIN_NCHANNELS=1`, and others), and disables AOT compile. These are read when cuBLAS and NCCL initialize — once a cuBLAS handle or an NCCL communicator exists, setting them has no effect.
3. **Global precision settings.** It disables TF32 and reduced-precision reductions, since both introduce non-deterministic rounding.

So the order is forced:

```text
select CUDA device
  → init_batch_invariance()     operators replaced, env set, TF32 off
  → initialize distributed      NCCL reads its determinism variables here
  → construct model             every layer picks up the replaced operators
  → forward
```

Initializing at first forward is not "late" — it is inert. The model has already been built from the original operators, and cuBLAS and NCCL have already fixed their configuration.

The same mechanism means it cannot be switched off in place: the enable path is idempotent with no counterpart, `torch.bmm` has been rebound, and the environment variables have already been consumed. Returning to the default path requires a fresh worker process. That is an upstream property, not a choice in this design.

## Delivery plan

Three parts, each independently reviewable. For each one, "After this part" describes what a user can actually observe once it lands — as distinct from what code changed.

### Part 1 — Seed contract

**Goal.** Make a request's random source survive the trip to the worker.

Today the seed does not. Plain `SamplingParams` from the caller gets normalized into a fresh `OmniDiffusionSamplingParams` and the seed is dropped in that conversion; the request object then fills in a random one. Both the initial-submission and streaming-update paths need fixing under BIC mode.

**After this part.** Pass `seed=1234` and the worker receives 1234. Pass nothing, or a generator instead of a seed, and you get an error naming the problem rather than a silent substitution. Two fresh processes with the same seed at `max_num_seqs=1` then produce identical output — the reference baseline Part 2 compares against.

### Part 2 — One certified configuration

**Goal.** Prove that on one narrow configuration, a fixed request produces bitwise-identical output alone at BS1 and inside a real BSN batch.

```text
pipeline    StableDiffusion3Pipeline
device      1x NVIDIA CUDA GPU, compute capability >= 8.0
parallelism TP=1 DP=1 PP=1 SP=1 CFG-parallel=1, VAE patch parallel=1
execution   eager (no compile, no CUDA graphs)
dtype       bfloat16
shape       fixed height, width, steps, dtype, output count
guidance    guidance_scale=1.0, no negative prompt
RNG         explicit integer seed per request
disabled    LoRA, cache backends, prompt-embed cache, CPU and layerwise
            offload, quantization, custom pipeline args, runner overrides
input       standard text prompt; externally injected latents/sigmas rejected
```

**After this part.** On this configuration, the same request under concurrency gives the same bytes every time regardless of batch slot or neighbors, with published evidence behind it. On anything else, enabling BIC fails at startup or submission naming what is unsupported — so you find out immediately instead of trusting a guarantee nobody measured. With BIC off, nothing changes.

**The controls are the point.** A passing positive result proves little on its own, so the evidence needs three legs: the forward matrix (BS1 vs real BS2/BS4, with batch-index, order, and neighbor-mutation perturbations across fresh processes); a **feature-off control** — if the matrix also passes with BIC disabled, the pipeline was already invariant and this work delivers nothing; and a component-isolation control, disabling one suspected contributor at a time so a residual difference can be attributed rather than explained away.

**Guard against false passes.** Every case asserts exactly one pipeline forward, the exact scheduled request IDs, and observed batch dimension N — otherwise a quietly serialized run passes as co-batched. Comparisons are on reloaded tensors with dtype, shape, and byte-level hashes, never on encoded image bytes. And describe results at the granularity measured: if a stage reaches equality by approximating BS1 rather than by genuine co-batched invariance, say so per stage instead of generalizing.

**Method constraint.** Measure first, then enable native BIC, then patch only the first operation evidence proves non-invariant. The closed attempt #4857 changed 33 files and got reproducibility by capping effective concurrency at one — the serialized shortcut, not a baseline.

### Part 3 — Downstream consumer

**Goal.** Let a training framework consume the capability without vLLM-Omni depending on it.

vLLM-Omni exposes a standard trajectory/logprob output interface; the training side consumes it. `OmniDiffusionSamplingParams` already declares `return_trajectory_latents`, `return_trajectory_decoded`, `trajectory_timesteps`, and `trajectory_latents` with no pipeline reading them — that is the natural seam.

**After this part.** A fixed-topology rollout and a short fixed-topology training run reproduce across two fresh jobs, feature defaulting off.

## Related work

- [vLLM batch invariance docs](https://docs.vllm.ai/en/latest/features/batch_invariance/)
- [vLLM reproducibility docs](https://docs.vllm.ai/en/latest/usage/reproducibility/)
- [VeRL PR #6572](https://github.com/verl-project/verl/pull/6572)
- [VeRL-Omni issue #203](https://github.com/verl-project/verl-omni/issues/203)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by ZJLi2013** [comment]
need add  **conv** in layer3, verified on dreamzero (with VAE blocks),  `init_batch_invariance()` touches conv only via `torch.backends.cudnn.conv.fp32_precision = "ieee"` — precision, not reduction order — and never sets `torch.backends.cudnn.deterministic` 

Measured on gfx942 (diffusion pipeline, single offline process, `max_num_seqs=1`, fixed input, no concurrency): the overridden operator classes are already bitwise stable run-to-run at the pipeline's real shapes, while the pipeline end-to-end is not, about **1e-2** diff. then with 
`cudnn.deterministic = True` makes e2e bitwise reproducible. while it cost "+61%" latency cost per inference.  should bee same on NV GPUs

just found #3435 has mentioned this clearly:  "Bit-wise Alignment: Provide an option or backend configuration to enforce deterministic CUDA operations (e.g., torch.backends.cudnn.deterministic = True and setting torch.use_deterministic_algorithms(True))"
